File: backend/peated/models/change.py
from sqlalchemy import BigInteger, Column, DateTime, Enum, ForeignKeyConstraint, PrimaryKeyConstraint, Text, text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import relationship

from peated.db.base_class import Base


class Change(Base):
    __tablename__ = "change"
    __table_args__ = (
        ForeignKeyConstraint(["created_by_id"], ["user.id"], name="change_created_by_id_user_id_fk"),
        PrimaryKeyConstraint("id", name="change_pkey"),
    )

    id = Column(BigInteger)
    object_id = Column(BigInteger, nullable=False)
    object_type = Column(
        Enum(
            "bottle",
            "edition",
            "entity",
            "tasting",
            "toast",
            "follow",
            "comment",
            name="object_type",
        ),
        nullable=False,
    )
    data = Column(JSONB, nullable=False, server_default=text("'{}'::jsonb"))
    created_at = Column(DateTime, nullable=False, server_default=text("now()"))
    created_by_id = Column(BigInteger, nullable=False)
    type = Column(
        Enum("add", "update", "delete", name="type"),
        nullable=False,
        server_default=text("'add'::type"),
    )
    display_name = Column(Text)

    created_by = relationship("User")

File: src/server/database/model/change.go
package model

import (
	"time"

	"gorm.io/datatypes"
)

const (
	ChangeTypeAdd    string = "add"
	ChangeTypeDelete string = "delete"
	ChangeTypeUpdate string = "update"
)

type ObjectType string

const (
	ObjectTypeBottle  ObjectType = "bottle"
	ObjectTypeComment ObjectType = "comment"
	ObjectTypeEntity  ObjectType = "entity"
	ObjectTypeTasting ObjectType = "tasting"
	ObjectTypeToast   ObjectType = "toast"
	ObjectTypeFollow  ObjectType = "follow"
)

type Change struct {
	ID          uint64         `gorm:"primaryKey"`
	ObjectID    uint64         `gorm:"column:object_id;not null" json:"objectId"`
	ObjectType  ObjectType     `gorm:"column:object_type;not null" json:"objectType"`
	Type        string         `gorm:"column:type;not null"`
	DisplayName *string        `gorm:"column:display_name" json:"displayName"`
	Data        datatypes.JSON `gorm:"type:jsonb;not null"`

	CreatedByID uint64    `gorm:"column:created_by_id;not null" json:"createdById"`
	CreatedAt   time.Time `gorm:"column:created_at;not null" json:"createdAt"`

	CreatedBy User
}

func (Change) TableName() string {
	return "change"
}

type Changes []*Change
